fem/transfer.cpp

```diff
@@ -548,117 +1342,550 @@ void L2ProjectionGridTransfer::L2ProjectionL2Space::ProlongateTranspose(
+      for (int i = 0; i < ir_ho->GetNPoints(); ++i)
+      {
+         const IntegrationPoint& ip_ho = ir_ho->IntPoint(i);
+         fe_ho.CalcShape(ip_ho, shape_ho);
+         el_tr_ho->SetIntPoint(&ip_ho);
+         MH_el += (shape_ho *= (el_tr_ho->Weight() * ip_ho.weight));
+      }
+      fes_ho.GetElementDofs(iho, dofs_ho);
+      M_H.AddElementVector(dofs_ho, MH_el);
+
+      // Similarly for LOR mass matrix: loop over every LOR element in each HO element
+      const FiniteElement& fe_lor = *fes_lor.GetFE(lor_els[0]);
+      int nedof_lor = fe_lor.GetDof();
+      Vector ML_el(nedof_lor);
+      Vector shape_lor(nedof_lor);
+      Array<int> dofs_lor(nedof_lor);
+      for (int iref = 0; iref < nref; ++iref)
+      {
+         int ilor = lor_els[iref];
+         ElementTransformation* el_tr = fes_lor.GetElementTransformation(ilor);
+
+         // to re-use coeff, use integration rule from coeff (higher than needed)
+         ML_el = 0.0;
+         int nqPts = ir.GetNPoints();
+         for (int i = 0; i < nqPts; ++i)
+         {
+            const IntegrationPoint& ip_lor = ir.IntPoint(i);
+            fe_lor.CalcShape(ip_lor, shape_lor);
+            el_tr->SetIntPoint(&ip_lor);
+            ML_el += (shape_lor *= (el_tr->Weight() * ip_lor.weight * qfunc[ilor*nqPts +
+                                                                                       i]));
+         }
+         fes_lor.GetElementDofs(ilor, dofs_lor);
+         ML_inv_ea.AddElementVector(dofs_lor, ML_el);
+      }
+   }
+   // DOF by DOF inverse of non-zero entries
+   LumpedMassInverse(ML_inv_ea);
+
+
+   // **************************
+   // mixed mass M_LH
+   // **************************
+   MixedMassEA(fes_ho, fes_lor, coeff, M_LH_ea, d_mt);
+
+   // Set ownership
+   M_LH_local_op = new H1SpaceMixedMassOperator(&fes_ho, &fes_lor, &ho2lor,
+                                                &M_LH_ea);
+
+   ML_inv_vea.reset(new H1SpaceLumpedMassOperator(&fes_ho, &fes_lor, ML_inv_ea));
+   M_LH_vea.reset(M_LH_local_op);
+   R_vea.reset(new ProductOperator(ML_inv_vea.get(), M_LH_vea.get(), false,
+                                   false));
+
+   Array<int> ess_tdof_list;  // leave empty
+   precon_vea.reset(new OperatorJacobiSmoother(M_H, ess_tdof_list));
+
+   TransposeOperator* R_veaT = new TransposeOperator(R_vea.get());
+   RTxM_LH_vea.reset(new ProductOperator(R_veaT, M_LH_vea.get(), false, false));
+
+   DeviceSetupPCG();
 }
 
-L2ProjectionGridTransfer::L2ProjectionH1Space::L2ProjectionH1Space(
-   const FiniteElementSpace& fes_ho_, const FiniteElementSpace& fes_lor_)
-   : L2Projection(fes_ho_, fes_lor_)
-{
-   std::unique_ptr<SparseMatrix> R_mat, M_LH_mat;
-   std::tie(R_mat, M_LH_mat) = ComputeSparseRAndM_LH();
+#ifdef MFEM_USE_MPI
+
+void L2ProjectionGridTransfer::L2ProjectionH1Space::DeviceL2ProjectionH1Space(
+   const ParFiniteElementSpace& pfes_ho, const ParFiniteElementSpace& pfes_lor,
+   const Coefficient* coeff_)
```

Comment:
It looks like there is a lot of code duplication between (a) the parallel and serial versions of these objects, and (b) the classes L2ProjectionH1Space and L2ProjectionL2Space. For maintainability, the duplication should be reduced as much as possible.